extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTools.java

```diff
@@ -220,21 +221,122 @@ public static void writeTable(@NotNull final Table sourceTable,
+                // layers and delete any existing backup files for these grouping columns
+                final String[] parquetColumnNames = Arrays.stream(groupingColumnNames)
+                        .map(writeInstructions::getParquetColumnNameFromColumnNameOrDefault)
+                        .toArray(String[]::new);
+                groupingColumnWritingInfoMap =
+                        groupingColumnInfoBuilderHelper(groupingColumnNames, parquetColumnNames, destFile);
+            }
+
+            // Store all the files for cleanup in case of exception
+            cleanupFiles.add(shadowDestFile);
+            if (groupingColumnWritingInfoMap != null) {
+                for (final ParquetTableWriter.GroupingColumnWritingInfo gcwi : groupingColumnWritingInfoMap.values()) {
+                    cleanupFiles.add(gcwi.destFile);
+                }
+            }
+
+            ParquetTableWriter.write(sourceTable, definition, writeInstructions, shadowDestFile.getPath(),
+                    Collections.emptyMap(), groupingColumnWritingInfoMap);
+
+            // Write to shadow files was successful
+            installShadowFile(destFile, shadowDestFile);
+            rollbackFiles.add(destFile);
+            if (groupingColumnWritingInfoMap != null) {
+                for (final ParquetTableWriter.GroupingColumnWritingInfo gfi : groupingColumnWritingInfoMap.values()) {
+                    final File groupingDestFile = gfi.metadataFilePath;
+                    final File shadowGroupingFile = gfi.destFile;
+                    installShadowFile(groupingDestFile, shadowGroupingFile);
+                    rollbackFiles.add(groupingDestFile);
+                }
+            }
         } catch (Exception e) {
+            for (final File file : rollbackFiles) {
+                rollbackFile(file);
+            }
+            for (final File file : cleanupFiles) {
+                file.delete();
+            }
             if (firstCreated != null) {
                 FileUtils.deleteRecursivelyOnNFS(firstCreated);
-            } else {
-                // noinspection ResultOfMethodCallIgnored
-                destFile.delete();
             }
-            throw e;
+            throw new UncheckedDeephavenException("Error writing table to " + destFile.getPath(), e);
+        }
+    }
+
+    private static File getShadowFile(File destFile) {
+        return new File(destFile.getParent(), ".NEW_" + destFile.getName());
+    }
+
+    @VisibleForTesting
+    static File getBackupFile(File destFile) {
+        return new File(destFile.getParent(), ".OLD_" + destFile.getName());
+    }
+
+    private static String minusParquetSuffix(@NotNull final String s) {
+        if (s.endsWith(PARQUET_FILE_EXTENSION)) {
+            return s.substring(0, s.length() - PARQUET_FILE_EXTENSION.length());
+        }
+        return s;
+    }
+
+    public static Function<String, String> defaultGroupingFileName(@NotNull final String path) {
```

Comment:
We probably need a new issue to nest these in a subdirectory for grouping info. It creates issues with flat partitioned layouts today.